DOCS-10418 (#28319)
DataDog/documentation: content/en/tracing/trace_collection/automatic_instrumentation/single-step-apm/_index.md

{{% /tab %}}

> {{% tab "Kubernetes (Preview)" %}}

Author: Fine by me!

{{< /collapse-content >}}

{{< collapse-content title="Installing with Helm" level="h4" >}}
Follow these steps to enable Single Step Instrumentation across your entire cluster with Helm. This automatically sends traces for all applications in the cluster that are written in supported languages. 

> **Note:** For instructions on how to restrict Single Step Instrumentation to specific namespace or pods, see [Advanced options](#advanced-options).

Reviewer: ```suggestion
**Note:** For instructions on how to configure Single Step Instrumentation for specific namespace or pods, see [Advanced options](#advanced-options).
```

{{% /tab %}}

{{% tab "Kubernetes (Agent v7.64+)" %}}

### Configuring instrumentation for namespaces and pods

> By default, Single Step Instrumentation will instrument all services in all namespaces in your cluster. Alternatively, you can create targeting blocks with the `targets` label to specify which workloads to instrument and what configurations to apply.

Author: Wanna take a first stab at what that "something" would entail and I can fold it in? :)

> By default, Single Step Instrumentation will instrument all services in all namespaces in your cluster. Alternatively, you can create targeting blocks with the `targets` label to specify which workloads to instrument and what configurations to apply.

Reviewer: I tried writing something, but for now it looks mostly like linking to the k8s docs with like one more sentence :) which is what you did, thanks!

Each target block has the following keys:

| Key             | Description |
|------------------|-------------|
| `name`            | The name of the target block. This has no effect on monitoring state and is used only as metadata. |
| `namespaceSelector` | The namespace(s) to instrument. Specify using either:<br> - `matchNames`: One or more namespace name(s). <br> - `matchLabels`: One or more namespace label(s). |
| `podSelector`     | The pod(s) to instrument. Specify using: <br> - `matchLabels`: One or more pod label(s). |

> | `ddTraceVersions` | The Datadog APM SDK version to use for each language. |

Reviewer: Can we also link `language` to some place where we describe supported language SDKs on this page (or elsewhere)?

Author: @stanistan Would this work? https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/dd_libraries/

Or I can link to the compatibility page: https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/single-step-apm/compatibility/?code-lang=java#tracer-libraries

Reviewer: I think the compatibility page is better here since the reader is in the details and the `dd_libaries` link will kick them out to a top level language setup.

https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/single-step-apm/compatibility/#tracer-libraries (removing the `?code-lang=java` in the url)

Reviewer: Made this change

| `ddTraceConfigs`  | APM SDK configs that allow setting Unified Service Tags, enabling Datadog products beyond tracing, and customizing other APM settings. [See full list of options][1]. |

The file you need to configure depends on how you enabled Single Step Instrumentation:
- If you enabled SSI with Datadog Operator, edit `datadog-agent.yaml`.
- If you enabled SSI with Helm, edit `datadog-values.yaml`.

#### Example configurations

Review the following examples demonstrating how to select specific services:

{{< collapse-content title="Example 1: Enable all namespaces except one" level="h4" >}}

This configuration:
- enables APM for all namespaces except the `jenkins` namespace. 
- instructs Datadog to instrument the Java applications with the default Java APM SDK and Python applications with `v.3.1.0` of the Python APM SDK.

{{< highlight yaml "hl_lines=4-10" >}}
   apm:  
     instrumentation:  
       enabled: true  
     disabledNamespaces:  
       - "jenkins"  
     targets:  
       - name: "all-remaining-services"  
         ddTraceVersions:  
           java: "default"  

>             python: "3.1.0"

Reviewer: ```suggestion
           python: "3.1.0"
```

- instructs Datadog to:
  - instrument services in `login-service` with the default version of the Java APM SDK.
  - instrument services in `billing-service` with `v3.1.0` of the Python APM SDK. 
- sets several Datadog environment variables to apply to this target.

{{< highlight yaml "hl_lines=4-28" >}}
  apm:
    instrumentation:
      enabled: true
    targets:
      - name: "login-service_namespace"
        namespaceSelector:
          matchNames:
            - "login-service"
        ddTraceVersions:
          java: "default"
        ddTraceConfigs:
          - name: "DD_SERVICE"
            value: "login-service"
          - name: "DD_ENV"
            value: "prod"
          - name: "DD_PROFILING_ENABLED"  ## profiling is enabled for all services in this namespace
            value: "auto"
      - name: "billing-service_apps"
        namespaceSelector:
          matchLabels:
            app: "billing-service"
        ddTraceVersions:
          python: "3.1.0"
        ddTraceConfigs:
          - name: "DD_SERVICE"
            value: "billing-service"
          - name: "DD_ENV"
            value: "prod
{{< /highlight >}}

{{< /collapse-content >}}

> {{< collapse-content title="Example 3: Instrument a subset of pods" level="h4" >}}

Reviewer: Instrument different workloads with different tracers.

The file you need to configure depends on how you enabled Single Step Instrumentation:
- If you enabled SSI with Datadog Operator, edit `datadog-agent.yaml`.
- If you enabled SSI with Helm, edit `datadog-values.yaml`.

#### Example configurations

Review the following examples demonstrating how to select specific services:

{{< collapse-content title="Example 1: Enable all namespaces except one" level="h4" >}}

This configuration:
- enables APM for all namespaces except the `jenkins` namespace. 
- instructs Datadog to instrument the Java applications with the default Java APM SDK and Python applications with `v.3.1.0` of the Python APM SDK.

{{< highlight yaml "hl_lines=4-10" >}}
   apm:  
     instrumentation:  
       enabled: true  
       disabledNamespaces:  
         - "jenkins"  
       targets:  
         - name: "all-remaining-services"  
           ddTraceVersions:  
             java: "default"  
             python: "3.1.0"
{{< /highlight >}}

{{< /collapse-content >}}

{{< collapse-content title="Example 2: Instrument a subset of namespaces, matching on names and labels" level="h4" >}}

This configuration creates two targets blocks:

> - The first block (nameed `login-service_namespace`):

Author: ```suggestion
- The first block (named `login-service_namespace`):
```

  - sets environment variables -- `DD_SERVICE` and `DD_ENV` -- for this target group.

{{< highlight yaml "hl_lines=4-28" >}}
  apm:
    instrumentation:
      enabled: true
      targets:
        - name: "login-service_namespace"
          namespaceSelector:
            matchNames:
              - "login-service"
          ddTraceVersions:
            java: "default"
          ddTraceConfigs:
            - name: "DD_SERVICE"
              value: "login-service"
            - name: "DD_ENV"
              value: "prod"
            - name: "DD_PROFILING_ENABLED"  ## profiling is enabled for all services in this namespace
              value: "auto"
        - name: "billing-service_apps"
          namespaceSelector:
            matchLabels:
              app: "billing-service"
          ddTraceVersions:
            python: "3.1.0"
          ddTraceConfigs:
            - name: "DD_SERVICE"
              value: "billing-service"
            - name: "DD_ENV"
              value: "prod
{{< /highlight >}}

{{< /collapse-content >}}

{{< collapse-content title="Example 3: Instrument different workloads with different tracers" level="h4" >}}

> This configuration:
> - enables APM for pods labeled:
>   - `app:db-user`, which run the `db-user` application.
>   - `webserver:routing`, which run the `request-router` application.
> - instructs Datadog to use the default versions of the Datadog Tracer SDKs.
> - sets several Datadog environment variables to apply to each target group.

Author: ```suggestion
This configuration does the following:
- enables APM for pods with the following labels:
  - `app:db-user`, on pods running the `db-user` application.
  - `webserver:routing`, on pods running the `request-router` application.
- instructs Datadog to use the default versions of the Datadog Tracer SDKs.
- sets several Datadog environment variables to apply to each target group.
```

{{< /highlight >}}

{{< /collapse-content >}}

{{< collapse-content title="Example 4: Instrument a pod within a namespace" level="h4" >}}

This configuration:
- enables APM for pods labeled `app:password-resolver` inside the `login-service` namespace.
- instructs Datadog to use the default version of the Datadog Java Tracer SDK.
- sets several Datadog environment variables to apply to this target.

{{< highlight yaml "hl_lines=4-28" >}}
   apm:
     instrumentation:
       enabled: true
       targets:
         - name: "login-service-namespace"
           namespaceSelector:
             matchNames:
               - "login-service"
           podSelector:
             matchLabels:
               app: "password-resolver"
           ddTraceVersions:
             java: "default"
           ddTraceConfigs:   
             - name: "DD_SERVICE"
               value: "password-resolver"
             - name: "DD_ENV"
               value: "prod"
             - name: "DD_PROFILING_ENABLED"  
               value: "auto"
{{< /highlight >}}

{{< /collapse-content >}}

{{< collapse-content title="Example 5: Instrument a subset of pods using <code>matchExpressions</code>" level="h4" >}}

> This configuration enables APM for pods that meet the following criteria:
> 
> - for all apps except for the ones that have either of the labels `app=app1` or `app=app2`

Author: ```suggestion
This configuration enables APM for all pods except those that have either of the labels `app=app1` or `app=app2`.
```

2. Restart the service.
{{% /tab %}}

{{% tab "Kubernetes (Preview)" %}}

> 

Reviewer: Added a sentence:

> As an alternative, or for a version of the agent that does not support workload selection, you can also disable pod mutation by adding a label to your pod.

2. Restart the service.
{{% /tab %}}

{{% tab "Kubernetes (Preview)" %}}

#### Use workload selection (recommended)

> With workload selection, you can enable and disable tracing for specific applications. [See configuration details here](#advanced-options).

Reviewer: I don't know if there's a way to link to the explicit panel that can open because when you go here it throws you back to the _top_ of the entire section and is a little confusing.
